janus_server/src/bin/aggregation_job_driver.rs

```diff
@@ -0,0 +1,1422 @@
+            // semaphore is closed, and we never close this semaphore.
+            let _ = sem.acquire().await.unwrap();
+
+            // Acquire some aggregation jobs which are ready to be stepped.
+            //
+            // We determine the maximum number of jobs to acquire based on the number of semaphore
+            // permits available, since we'd like to start processing any acquired jobs immediately
+            // to avoid potentially timing out while waiting on _other_ jobs to finish being
+            // stepped. This is racy given that workers may complete (and relinquish their permits)
+            // concurrently with us acquiring jobs; but that's OK, since this can only make us
+            // underestimate the number of jobs we can acquire, and underestimation is acceptable
+            // (we'll pick up any additional jobs on the next iteration of this loop). We can't
+            // overestimate since this task is the only place that permits are acquired.
+
+            // TODO(brandon): only acquire jobs whose batch units have not already been collected (probably by modifying acquire_incomplete_aggregation_jobs)
+            let max_acquire_count = sem.available_permits();
+            info!(max_acquire_count, "Acquiring aggregation jobs");
+            let acquired_jobs = self
+                .datastore
+                .run_tx(|tx| {
+                    let lease_duration = self.aggregation_worker_lease_duration;
+                    Box::pin(async move {
+                        tx.acquire_incomplete_aggregation_jobs(lease_duration, max_acquire_count)
+                            .await
+                    })
+                })
+                .await;
+            let acquired_jobs = match acquired_jobs {
+                Ok(acquired_jobs) => acquired_jobs,
+                Err(err) => {
+                    error!(?err, "Couldn't acquire aggregation jobs");
+                    // Go ahead and step job discovery delay in this error case to ensure we don't
+                    // tightly loop running transactions that will fail without any delay.
+                    job_discovery_delay = self.step_job_discovery_delay(job_discovery_delay);
+                    continue;
+                }
+            };
+            if acquired_jobs.is_empty() {
+                debug!("No aggregation jobs available");
+                job_discovery_delay = self.step_job_discovery_delay(job_discovery_delay);
+                continue;
+            }
+            info!(
+                acquired_job_count = acquired_jobs.len(),
+                "Acquired aggregation jobs"
+            );
+
+            // Start up tasks for each acquired aggregation job.
+            job_discovery_delay = Duration::ZERO;
+            for (task_id, vdaf, aggregation_job_id, lease_expiry) in acquired_jobs {
+                info!(
+                    ?task_id,
+                    ?aggregation_job_id,
+                    ?lease_expiry,
```

Comment:
👍🏻 Note I used `.instrument()` rather than the `#[tracing::instrument]` macro on `step_aggregation_job` because I wanted to scope the span to include the error-handling code.